### PR TITLE
Core: Unique outputDir/cacheDir for each configDir

### LINKED ISSUE #24082: [Bug]: Running multiple dev server concurrently with composition results in applications failing at runtime with 404 Not found manager-bundle.js
### Describe the bug

Currently running multiple dev server at once with `concurrently`, `@storybook/react-vite` framework and a composition setup:
- main app: port 6006 with use composition and display components from applications at port 6007 and 6008
  - refs composition is configured and is working
- app 1: port 6007
- app 2: port 6008

When accessing main application at port 6006, it works, all composed components will load as well without issues.

Issue occurs when trying to load port 6007 or 6008, we are seeing a white page and receiving a 404 Not found error on files:
- port 6007: http://localhost:6007/sb-addons/storybook-APP1-10/manager-bundle.js
- port 6008: http://localhost:6008/sb-addons/storybook-APP2-11/manager-bundle.js

image

Looks like when running multiple dev server concurrently the storybook cache gets overwritten by each app in a shared cache folder at `node_modules/.cache/storybook/public/sb-addons`. Because, when running the terminals on 3 different repositories everything work fine.

Is there any way to configure the storybook cache folder location per application/configuration or deactivate such cache entirely?

### To Reproduce

Package.json
```typrescript
// deps
"@storybook/addon-actions": "^7.4.0",
"@storybook/addon-essentials": "^7.4.0",
"@storybook/addon-interactions": "^7.4.0",
"@storybook/addon-links": "^7.4.0",
"@storybook/addon-viewport": "^7.4.0",
"@storybook/addons": "^7.4.0",
"@storybook/jest": "^0.2.2",
"@storybook/node-logger": "^7.4.0",
"@storybook/preset-create-react-app": "^7.4.0",
"@storybook/react": "^7.4.0",
"@storybook/react-vite": "^7.4.0",
"@storybook/test-runner": "^0.13.0",
"@storybook/testing-library": "^0.2.0",

 // dev server scripts
"start-storybook": "concurrently -k -s first -n \"APP1,APP2,MAIN\" -c \"magenta,blue,green\" \"npm run start-storybook:APP1\" \"npm run start-storybook:APP2\" \"wait-on tcp:6007 && wait-on tcp:6008 && npm run start-storybook:MAIN_APP\"",
"start-storybook:APP1": "storybook dev -c .storybook/apps/app1 -p 6007 --no-open",
"start-storybook:APP2": "storybook dev -c .storybook/apps/app2-p 6008 --no-open",
"start-storybook:MAIN_APP": "storybook dev -c .storybook/apps/main-p 6006 --no-open",
```

Refs on main application configuration:
```
refs: (config, { configType }) => {
    if (configType === 'DEVELOPMENT') {
      return {
        'APP1': {
          title: 'APP1',
          url: 'http://localhost:6007',
        },
        'APP2': {
          title: 'APP2',
          url: 'http://localhost:6008',
        },
      };
    }
    return {
      'APP1': {
        title: 'APP1',
        url: 'PROD_URL',
      },
      'APP2': {
        title: 'APP2',
        url: 'PROD_URL',
      },
    };
  },
```

### System

```bash
Environment Info:

  System:
    OS: Linux 5.15 Ubuntu 22.04.2 LTS 22.04.2 LTS (Jammy Jellyfish)
    CPU: (10) x64 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz
  Binaries:
    Node: 20.4.0 - ~/.nvm/versions/node/v20.4.0/bin/node
    npm: 9.7.2 - ~/.nvm/versions/node/v20.4.0/bin/npm
  npmPackages:
    @storybook/addon-actions: ^7.4.0 => 7.4.0 
    @storybook/addon-essentials: ^7.4.0 => 7.4.0 
    @storybook/addon-interactions: ^7.4.0 => 7.4.0 
    @storybook/addon-links: ^7.4.0 => 7.4.0 
    @storybook/addon-viewport: ^7.4.0 => 7.4.0 
    @storybook/addons: ^7.4.0 => 7.4.0 
    @storybook/jest: ^0.2.2 => 0.2.2 
    @storybook/node-logger: ^7.4.0 => 7.4.0 
    @storybook/preset-create-react-app: ^7.4.0 => 7.4.0 
    @storybook/react: ^7.4.0 => 7.4.0 
    @storybook/react-vite: ^7.4.0 => 7.4.0 
    @storybook/test-runner: ^0.13.0 => 0.13.0 
    @storybook/testing-library: ^0.2.0 => 0.2.0
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ndelangen**
Closes #24082

I think this will also fix this issue: #21510

## What I did

I ensured that storybook when using a cache directory as an output directory, will use a unique sub-dir per storybook configDir.

I also cleaned up a bit of terminal output, there were absolute paths in there, which feels noisy.
It also printed out a message about copying over things from node_modules, which I doubt is useful information to the user, as it was storybook's internal pre-bundled stuff that was getting copied)

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [ ] unit tests
- [x] integration tests
- [x] end-to-end tests

#### Manual testing

- Making a static build of any storybook, and check if it work.
- Provide a manual outputDir via a CLI flag and see if it works

- Start a dev sandbox
- check the `node_modules/.cache` directory. It's contents should include:
  Screenshot 2023-12-19 at 12 06 09

The long hash is unique per storybook configDir. It does not leak anything personally identifiable.
Here's the code creating it:
https://github.com/storybookjs/storybook/blob/3b94da1a9f4e5942dae067f277c2a4672271cf2e/code/lib/core-server/src/build-dev.ts#L60

- check if the dev version works as expected.

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

### CODE REVIEW

File: code/lib/core-server/src/utils/copy-all-static-files.ts [hunk 1]
```diff
@@ -11,11 +11,17 @@ export async function copyAllStaticFiles(staticDirs: any[] | undefined, outputDi
       staticDirs.map(async (dir) => {
         try {
           const { staticDir, staticPath, targetDir } = await parseStaticDir(dir);
-          const targetPath = path.join(outputDir, targetDir);
-          logger.info(chalk`=> Copying static files: {cyan ${staticDir}} => {cyan ${targetDir}}`);
+          const targetPath = join(outputDir, targetDir);
+          if (!staticDir.includes('node_modules')) {
```
Comment (shilman) [inline code/lib/core-server/src/utils/copy-all-static-files.ts:15]: This line could use a comment explaining why

File: code/lib/core-server/src/utils/copy-all-static-files.ts [hunk 2]
```diff
@@ -49,13 +55,21 @@ export async function copyAllStaticFilesRelativeToMain(
...
-    const skipPaths = ['index.html', 'iframe.html'].map((f) => path.join(targetPath, f));
-    logger.info(chalk`=> Copying static files: {cyan ${from}} at {cyan ${targetPath}}`);
+    const targetPath = join(outputDir, to);
+    const skipPaths = ['index.html', 'iframe.html'].map((f) => join(targetPath, f));
+    if (!from.includes('node_modules')) {
+      logger.info(
+        chalk`=> Copying static files: {cyan ${print(from)}} at {cyan ${print(targetPath)}}`
+      );
+    }
+    // logger.info(chalk`=> Copying static files: {cyan ${from}} at {cyan ${targetPath}}`);
```
Comment (shilman) [inline code/lib/core-server/src/utils/copy-all-static-files.ts:65]:
```suggestion
```

**Review by shilman — CHANGES_REQUESTED** [review]
Looks good!!! Minor changes

**Review by shilman — APPROVED** [review]

**Comment by domneo** [comment]
Will this be patched in V7 as well?

**Comment by shilman** [comment]
@domneo Nope. Please upgrade to the latest prerelease to use it. We're expecting a stable release in early 2024:

```
npx storybook@next upgrade --prerelease
```

**Comment by bobleap** [comment]
> Will this be patched in V7 as well?
+1

**Pull request merged**